apps/src/sites/studio/pages/levelbuilder.js

```diff
@@ -52,9 +53,30 @@ window.levelbuilder.copySelectedBlockToClipboard = function () {
       Blockly.Xml.blockToDom(Blockly.selected)
     );
     copyToClipboard(str);
+    localStorage.setItem('blockXml', str);
   }
 };
 
+window.levelbuilder.pasteBlocksToWorkspace = function () {
+  let str = localStorage.getItem('blockXml');
+
+  if (str.startsWith('<xml') && str.endsWith('</xml>')) {
+    // If an entire workspace has been copied, clear the current workspace.
+    Blockly.mainBlockSpace.clear();
```

Comment:
Could this produce some unexpected behavior (eg, unintentionally wiping unsaved work) if used incorrectly? I see some warning text in the haml file, but didn't see it displayed in the videos in your description?

Reply:
Ah, apologies. I added that text after capturing the videos:

![image](https://user-images.githubusercontent.com/43474485/234352002-c17871d2-80a5-43ce-bc42-25c1daa9a080.png)

It would wipe saved work if the levelbuilder (or engineer) using it also ran the new code to trigger a save. 

An example benefit of this behavior is that one could take a project on production and quickly copy the entire thing into the start blocks for a level.
It would probably be most problematic for a levelbuilder that forgot what they had most recently copied. Refreshing the page would restore the last saved blocks in this case.

Reply:
> An example benefit of this behavior is that one could take a project on production and quickly copy the entire thing into the start blocks for a level.

Is that not possible to do (if you're within the same lab) currently via keyboard shortcuts? Sorry, just trying to build out my understanding here a bit. I thought the main benefit here would be that you can go between labs that are on different versions of blockly?

I guess my hesitation is that I feel like this introduces sort of another levelbuilder feature that "kind of works, but only if you have some specific knowledge of how to use it", and doesn't super intuitively work with other ways of copying/pasting we support. The benefits probably outweighs this consideration, but curious for your thoughts.

Reply:
You can't copy or paste a full workspace worth of blocks using any keyboard shortcuts, no. 

I think there are multiple benefits. First, as you said, going between different Blockly versions is huge, especially for engineers, but also for curriculum writers working in Poetry/Time Capsule levels. Second, being able to quickly/cleanly copy a full workspace is something that would usually require jumping into level edit pages to pull off.

I do agree that this requires specific knowledge to use, similar to the other existing buttons. Those buttons also copy differently from the other methods you describe.

Reply:
If you didn't clear the workspace before pasting, is it straightforward for levelbuilders to delete the unwanted/duplicate blocks? I was thinking of a case when a levelbuilder may already have desirable blocks in a workspace and wants to paste a full workspace without removing current blocks.

Reply:
Great question, @fisher-alice! It is is absolutely not straightforward. Examples:
* Blocks may be configured to be undeleteable, which would require using the console to fix. `when run` is usually an example of this and would almost always be present.
* If the lab doesn't save block X/Y values, the pasted blocks could be placed directly over existing ones which might make it hard to even recognize that they are there (imagine pasting the same workspace blocks multiple times)
* If multiple function definition blocks with the same/similar names are in the same workspace, it could be tricky to determine which one shouldn't be there. Some functions are specified as not user-created and are not deleteable.
* CDO Blockly blocks may be invisible, e.g. behavior definitions. If pasted, they could be hard to find.

Reply:
Would adding an additional button clutter menu too much? Something like 'Clear and copy new workspace'...